src/SparseMFModel.cpp

```diff
@@ -28,333 +62,459 @@ void SparseMFModel::initialize_weights(uint64_t users, uint64_t items, uint64_t
+        std::get<1>(item_model).push_back(item_weight);
+      }
+      item_models[item_id] = item_model;
     }
+  }
+
+  /**
+   * userId : 0 to minibatch_size
+   */
+  FEATURE_TYPE SparseMFModel::predict(uint32_t userId, uint32_t itemId) const {
+    FEATURE_TYPE user_bias = std::get<1>(user_models[userId]);
+    FEATURE_TYPE item_bias = item_models[itemId].first;
+
+    FEATURE_TYPE res = global_bias_ + user_bias + item_bias;
+
+    for (uint32_t i = 0; i < nfactors_; ++i) {
+      res +=
+          std::get<2>(user_models[userId])[i] * item_models[itemId].second[i];
+#ifdef DEBUG
+      if (std::isnan(res) || std::isinf(res)) {
+        std::cout << "userId: " << userId << " itemId: " << itemId
+                  << " get_user_weights(userId, i): "
+                  << get_user_weights(userId, i)
+                  << " get_item_weights(itemId, i): "
+                  << get_item_weights(itemId, i) << std::endl;
+        throw std::runtime_error("nan error in predict");
+      }
 #endif
+    }
+    return res;
   }
-  return res;
-}
 
-std::unique_ptr<ModelGradient> SparseMFModel::minibatch_grad(
-            const SparseDataset& dataset,
-            const Configuration& config,
-            uint64_t base_user) {
-  FEATURE_TYPE learning_rate = config.get_learning_rate();
-  auto gradient = std::make_unique<MFSparseGradient>();
-  std::vector<FEATURE_TYPE> item_weights_grad_map[17770];
-  std::vector<int> item_weights_lst;
-  double training_rmse = 0;
-  uint64_t training_rmse_count = 0;
-
-  // iterate all pairs user rating
-  for (uint64_t user_from_0 = 0; user_from_0 < dataset.data_.size(); ++user_from_0) {
-    std::vector<FEATURE_TYPE> user_weights_grad(NUM_FACTORS);
-    uint64_t real_user_id = base_user + user_from_0;
-
-    // we have to populate this value in case this user doesn't have any ratings
-    // XXX we should probably optimize this
-    gradient->users_bias_grad[real_user_id] = 0;
-    for (uint64_t j = 0; j < dataset.data_[user_from_0].size(); ++j) {
-      // first user matches the model in user_models[0]
-      uint64_t itemId = dataset.data_[user_from_0][j].first;
-      FEATURE_TYPE rating = dataset.data_[user_from_0][j].second;
-
-      FEATURE_TYPE pred = predict(user_from_0, itemId);
-
-      FEATURE_TYPE error = rating - pred;
-      training_rmse += error * error;
-      training_rmse_count++;
-
-      // compute gradient for user bias
-      FEATURE_TYPE& user_bias = std::get<1>(user_models[user_from_0]);
-      float delta = learning_rate * (error - user_bias_reg_ * user_bias);
-      gradient->users_bias_grad[real_user_id] += delta;
-      user_bias += delta;
-
-
-      // compute gradient for item bias
-      FEATURE_TYPE& item_bias = item_models[itemId].first;
-      delta = learning_rate * (error - item_bias_reg_ * item_bias);
-      gradient->items_bias_grad[itemId] += delta;
-      item_bias += delta;
+  std::unique_ptr<ModelGradient> SparseMFModel::minibatch_grad(
```

Comment:
Many of the changes here are most likely due to blindly using clang-format blindly.

Reply:
I'll double back for this later.

Reply:
Fixed